evm-ext/src/events.rs

```diff
@@ -1,15 +1,68 @@
 use {
     super::{EVM_NATIVE_ADDRESS, EVM_NATIVE_MODULE},
-    alloy::primitives::{Log, LogData, B256},
+    alloy::primitives::{Log, LogData, B256, U256},
     move_core_types::{
+        account_address::AccountAddress,
         ident_str,
         language_storage::StructTag,
         value::{MoveStructLayout, MoveTypeLayout, MoveValue},
     },
     moved_shared::primitives::ToEthAddress,
-    std::sync::LazyLock,
+    std::{cell::RefCell, sync::LazyLock},
 };
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EthTransfer {
+    pub from: AccountAddress,
+    pub to: AccountAddress,
+    pub amount: U256,
+}
+
+pub trait EthTransferLog {
+    fn append_transfers(&self, transfers: Vec<EthTransfer>);
+    fn take_transfers(&self) -> Vec<EthTransfer>;
+    fn add_tx_origin(&self, address: AccountAddress, amount: U256);
+    fn take_origins(&self) -> Vec<(AccountAddress, U256)>;
+}
+
+#[derive(Debug, Default)]
+pub struct EthTransfersLogger {
+    transfers: RefCell<Vec<EthTransfer>>,
+    origins: RefCell<Vec<(AccountAddress, U256)>>,
+}
+
+impl EthTransferLog for EthTransfersLogger {
+    fn append_transfers(&self, mut transfers: Vec<EthTransfer>) {
+        self.transfers.borrow_mut().append(&mut transfers);
+    }
+
+    fn take_transfers(&self) -> Vec<EthTransfer> {
+        self.transfers.take()
+    }
+
+    fn add_tx_origin(&self, address: AccountAddress, amount: U256) {
+        self.origins.borrow_mut().push((address, amount));
+    }
+
+    fn take_origins(&self) -> Vec<(AccountAddress, U256)> {
+        self.origins.take()
+    }
+}
+
+impl EthTransferLog for () {
```

Comment:
I feel maybe this impl is better under `mod test_doubles` like other similar mock-ish impls we have?

Reply:
Right, and any other simulation like gas estimation, I believe. Makes sense.

Reply:
Well gas estimation actually does use the real implementation because the extra transfers that happen on the Move side to mirror the EVM cost gas and therefore need to be accounted for in the gas estimation.